otherlibs/systhreads/thread.ml

```diff
@@ -19,6 +19,126 @@
 
 [@@@ocaml.flambda_o3]
 
+(* Implementation copied from [Domain.DLS], modulo thread-safety concerns. *)
+module TLS = struct
+  module Obj_opt : sig @@ portable
+    type t
+    val none : t
+    val some : 'a -> t
+    val is_some : t -> bool
+    val unsafe_get : t -> 'a
+  end = struct
+    type t = Obj.t
+    let none = Obj.magic_portable (Obj.repr (ref 0))
+    let some v = Obj.repr v
+    let is_some obj = (obj != Obj.magic_uncontended none)
+    let unsafe_get obj = Obj.obj obj
+  end
+
+  type tls_state = Obj_opt.t array
+
+  external get_tls_state
+    : unit -> tls_state @@ portable = "caml_thread_get_state"
+  external set_tls_state
+    : tls_state -> unit @@ portable = "caml_thread_set_state"
+
+  let init () =
+    let state = Array.make 8 (Obj.magic_uncontended Obj_opt.none) in
```

Comment:
This is the standard approach &mdash; use a power of two sized array.  A minor issue with this is that an array whose size is a power of two ends up being power of two _plus one_ sized block.  And the problem with that is that it makes false sharing more likely than if the block size would be a power of two.  So, what I'd recommend is to use e.g. 7 here (i.e. power of two _minus one_) and then when growing the array, compute the new size as `old_size * 2 + 1`.  This way the memory block will have a power of two size and for small such blocks the allocator should keep them aligned in the major heap and eliminate possibility of false sharing.

Reply:
Good point, swapped both